### ISSUE TITLE
[Question] Does DQN copy `running_mean` and `running_var` of batch norm layers to the target network?

### ISSUE THREAD

**Issue by honglu2875** [DLR-RM/stable-baselines3#996]
I have been spending quite some time reading the codes here and I have been learning quite a lot so far. I got a small question when I backtrack some codes to find out why I got some unstable agents.

https://github.com/DLR-RM/stable-baselines3/blob/b1cc15970a40c86b26a247fab5783e025bfe3da1/stable_baselines3/dqn/dqn.py#L172

This line above seems to be the place where the target network is updated. But if there are batch norm layers in the network, we have some extra variables `running_mean` and `running_var` which are not contained in `Module.parameters()`. We know these are slightly different than those SGD updated parameters and they just keep tracking batch mean/var actively unless `.eval()` is called. 

It doesn't look like they are copied during target network update. But I believe the DDQN paper wants the target network to be a "periodic copy" of the original network which seems to imply that everything should be copied and stayed the same for certain amount of steps. 

Or are those `running_mean` and `running_var` updated somewhere else in the code base?

**Comment by Miffyli** [comment]
Somewhat related: #535

Good catch. I can not say for if the params are copied or not (as you said; probably not), however we do not have batch-norm layers in the default networks/policies in SB3. The default networks are from the old era before people started putting norm-layers everywhere :). I am not 100% sure how to approach this, though, unless there is a convenient pytorch way of getting all parameters like these from the model as well.

**Comment by honglu2875** [comment]
Gotcha! I can do some more searches and run some tests when I have time to see if batch norm layers are considered.

For my own codes (under the same variable naming), when I realized this problem I inserted these codes
```python
modules = []
modules_target = []
for module, module_target in zip(self.q_net.modules(), self.q_net_target.modules()):
    for key in ['running_mean', 'running_var']:
        if hasattr(module, key):
            modules.append(getattr(module, key))
            modules_target.append(getattr(module_target, key))
polyak_update(modules, modules_target, self.tau)
```
I assume the names `running_mean`, `running_var` will not be given to stuff unrelated to batch norm layers. But maybe this is still a bit adhoc...

**Comment by araffin** [comment]
Hello,
thanks for reporting the issue.
I also thought the batch norm states where included in the parameters, but actually no... See https://discuss.pytorch.org/t/batch-norm-parameters-not-included-in-model-parameters/10265/2

the idea behind is that batch norm stats are updated by doing forward pass, not by the optimizer.
we indeed need an additional helper to fix this issue and we would welcome a PR ;)

**Comment by araffin** [comment]
Note: one fast way to retrieve all batch norm params is like that:
```python
# Include both online net and target, need to be filtered to each separately
batch_norm_params = [name for name in model.policy.state_dict().keys() if "running_" in name]
```

**Comment by honglu2875** [comment]
> Note: one fast way to retrieve all batch norm params is like that:
> 
> ```python
> # Include both online net and target, need to be filtered to each separately
> batch_norm_params = [name for name in model.policy.state_dict().keys() if "running_" in name]
> ```

That looks much nicer! Looks like `state_dict` is by definition the actual set of all learnable parameters. What about we simply collect all items in `state_dict` that are tensors, and then scale them using `polyak_update`?
